src/modules/map/components/CompareSlider.vue

```diff
@@ -1,53 +1,109 @@
-const props = defineProps({
-    clientWidth: {
-        type: Number,
-        default: window.innerWidth,
-    },
-})
+const getMap = inject('getMap')
 
-const { clientWidth } = toRefs(props)
+const preComposeKey = ref(null)
+const postComposeKey = ref(null)
+const compareSliderOffset = ref(0)
 
 const compareRatio = ref(-0.5)
-
 const store = useStore()
 const storeCompareRatio = computed(() => store.state.ui.compareRatio)
+const clientWidth = computed(() => store.state.ui.width)
 const compareSliderPosition = computed(() => {
     return {
         left: compareRatio.value * 100 + '%',
     }
 })
+const visibleLayerOnTop = computed(() => store.getters.visibleLayerOnTop)
 
 watch(storeCompareRatio, (newValue) => {
     compareRatio.value = newValue
+    slice()
+})
+
+watch(visibleLayerOnTop, () => {
+    nextTick(slice())
 })
 
 onMounted(() => {
     compareRatio.value = storeCompareRatio.value
+    slice()
 })
 
-function grabSlider() {
+function slice() {
+    if (preComposeKey.value != null && postComposeKey.value != null) {
+        unByKey(preComposeKey.value)
+        unByKey(postComposeKey.value)
+        preComposeKey.value = null
+        postComposeKey.value = null
+    }
+    const aTopLayerEXist = getMap()
+        ?.getAllLayers()
+        .find((layer) => {
+            return visibleLayerOnTop.value && layer.get('id') === visibleLayerOnTop.value.getID()
+        })
+    if (aTopLayerEXist) {
+        preComposeKey.value = aTopLayerEXist.on('prerender', onPreCompose)
+        postComposeKey.value = aTopLayerEXist.on('postrender', onPostCompose)
+    }
+    getMap().render()
+}
+
+function onPreCompose(event) {
+    const ctx = event.context
+    // the offset is there to ensure we get to the slider line, and not the border of the element
+    var width = ctx.canvas.width
+    if (compareRatio.value < 1.0 && compareRatio.value > 0.0) {
+        width = ctx.canvas.width * compareRatio.value + 20
+    }
+
+    ctx.save()
+    ctx.beginPath()
+    ctx.rect(0, 0, width, ctx.canvas.height)
+    ctx.clip()
+}
+
+function onPostCompose(event) {
+    event.context.restore()
+    getMap().render()
+}
+
+function grabSlider(event) {
     window.addEventListener('mousemove', listenToMouseMove, { passive: true })
     window.addEventListener('touchmove', listenToMouseMove, { passive: true })
     window.addEventListener('mouseup', releaseSlider, { passive: true })
     window.addEventListener('touchend', releaseSlider, { passive: true })
+    if (event.type === 'touchstart') {
+        compareSliderOffset.value =
+            event.touches[0].clientX - compareRatio.value * clientWidth.value
+    } else {
+        compareSliderOffset.value = event.clientX - compareRatio.value * clientWidth.value
+    }
 }
 
 function listenToMouseMove(event) {
-    const currentPosition = event.type === 'touchmove' ? event.touches[0].pageX : event.pageX
-    compareRatio.value = round(currentPosition / clientWidth.value, 2)
+    var currentPosition
```

Comment:
don't use `var` keyword anymore (that's old school JS), use `let` instead

Reply:
The issue with this is that `let` and `const` are both scoped, which means if I want to use let, I have to repeat the code twice within the if - else

We have three options here

1. Revert to a `const` with a ternary operator
2. Use `var`
3. add the code that use the variable in both the if and the else

Reply:
if you use `let` that is the same thing as ` var`. `var` was the old way of describing a variable, without telling if it would later change or not. Hence the `let` and `const` distinction.

If you write the exact same code as there, but replace `var` with `let` it will function the same. I don't see the problem

Reply:
sorry, I was a bit dumb

let is indeed scoped, but not in the way I thought. It does indeed work like this